test/Microsoft.AspNetCore.Blazor.Test/TreeParameterStateTest.cs

```diff
@@ -0,0 +1,419 @@
+                Assert.Same(states[1].Component, match.FromProvider);
+            });
+        }
+
+        [Fact]
+        public void FindTreeParameters_IfHasMultipleMatchesInAncestors_ReturnsMatches()
+        {
+            // Arrange
+            var states = CreateAncestry(
+                new ComponentWithNoParams(),
+                CreateProvider(new TreeValue2()),
+                new ComponentWithNoParams(),
+                CreateProvider(new TreeValue1()),
+                new ComponentWithTreeParams());
+
+            // Act
+            var result = TreeParameterState.FindTreeParameters(states.Last());
+
+            // Assert
+            Assert.Collection(result.OrderBy(x => x.LocalName),
+                match => {
+                    Assert.Equal(nameof(ComponentWithTreeParams.TreeParam1), match.LocalName);
+                    Assert.Same(states[3].Component, match.FromProvider);
+                },
+                match => {
+                    Assert.Equal(nameof(ComponentWithTreeParams.TreeParam2), match.LocalName);
+                    Assert.Same(states[1].Component, match.FromProvider);
+                });
+        }
+
+        [Fact]
+        public void FindTreeParameters_InheritedParameters_ReturnsMatches()
+        {
+            // Arrange
+            var states = CreateAncestry(
+                CreateProvider(new TreeValue1()),
+                CreateProvider(new TreeValue3()),
+                new ComponentWithInheritedTreeParams());
+
+            // Act
+            var result = TreeParameterState.FindTreeParameters(states.Last());
+
+            // Assert
+            Assert.Collection(result.OrderBy(x => x.LocalName),
+                match => {
+                    Assert.Equal(nameof(ComponentWithTreeParams.TreeParam1), match.LocalName);
+                    Assert.Same(states[0].Component, match.FromProvider);
+                },
+                match => {
+                    Assert.Equal(nameof(ComponentWithInheritedTreeParams.TreeParam3), match.LocalName);
+                    Assert.Same(states[1].Component, match.FromProvider);
+                });
+        }
+
+        [Fact]
+        public void FindTreeParameters_ComponentRequestsBaseType_ReturnsMatches()
+        {
+            // Arrange
+            var states = CreateAncestry(
+                CreateProvider(new TreeValueDerivedClass()),
+                new ComponentWithGenericTreeParam<TreeValueBaseClass>());
+
+            // Act
+            var result = TreeParameterState.FindTreeParameters(states.Last());
+
+            // Assert
+            Assert.Collection(result, match => {
+                Assert.Equal(nameof(ComponentWithGenericTreeParam<object>.LocalName), match.LocalName);
+                Assert.Same(states[0].Component, match.FromProvider);
+            });
+        }
+
+        [Fact]
+        public void FindTreeParameters_ComponentRequestsImplementedInterface_ReturnsMatches()
```

Comment:
I think there's a decision point here about the combination of `is` based on the declared type of `T` for some `Provider<T>`. Generally you do one of these things or the other.

The pairings I expect to see are:

**declared type + `typeof(T) == ParameterType`** => Type is used as a 'key' or identity, rather then representing a set of optional policies/protocols

**runtime type + `ParameterType.IsAssignableFrom(Value.GetType())`** => Value can implement a set of protocols and binding is flexible

I expect the reason why it is like it is that you didn't have the value available in all cases in `CanSupplyValue` OR because you wanted the binding between to ancestor and parent to stay stable when the same set of components renders?

Reply:
Another consideration that I can see you thought of is `typed null`. Maybe what's here is great, I just want to talk a little more about the rationale.

Reply:
> Yes, this is an invariant I'm relying on to make the updates cheap. If this were not the case, then whenever any Provider value changes, we have to reconsider every possible descendant consumer to see whether the new runtime type matches, not just the ones that are currently matched to its value.

OK then it makes sense that we need a declared type. We'll just have to call this out clearly. My experience with similar features in model binding is that developers generally expect any system they use to have the maximum theoretical flexibility - but I think we have good reasons to make this based on declared typed. Ultimately we provide value by sometimes designing systems that limit flexibility where it's important.

----

> Do you have examples of any scenarios where doing it based on declared type would be problematic/inconvenient for the developer?

No, I don't think it's problematic, since the binding to a type is static. 

I was trying to rationalize this based on systems like WPF's static resources, or `IPropertyOwner` from VS's editor APIs where the type is used as a key. 

The only case that would become problematic is if you wanted to bind values to a common framework type like `IEquatable` or `IConvertable`. I think our guidance would be "don't do that". For instance with this system, it would be possible to define a cascading parameter of type `IComponent` and then bind to the nearest ancestor that is a component.  These things are of somewhat minimal utility, and somewhat average level of wierdness. None of it feels broken.

What this does allow you to do is treat a value you provide as a feature collection, but exporting a bunch of common interfaces. You could image defining a control system based on capability, and exporting commands. Consider something similar to [UIA control patterns](https://docs.microsoft.com/en-us/dotnet/framework/ui-automation/ui-automation-control-patterns-overview) - now you have bubbling commands 😁 

It seems really powerful as it is defined.